crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,331 @@
+use std::{
+    collections::btree_map::Range,
+    convert::TryInto,
+    io::{self, Read, Write},
+    net::{IpAddr, Ipv4Addr, SocketAddr, SocketAddrV4},
+    ops::RangeInclusive,
+    time::Duration,
+};
+
+use boringtun::noise;
+use pnet_packet::{
+    ip::{self, IpNextHeaderProtocols},
+    ipv4::{self, Ipv4Flags, Ipv4Packet, MutableIpv4Packet},
+    udp::{self, MutableUdpPacket, UdpPacket},
+    Packet,
+};
+use pqcrypto_kyber::kyber768;
+use pqcrypto_traits::kem::{Ciphertext, PublicKey, SharedSecret};
+use rand::{prelude::Distribution, SeedableRng};
+use tokio::net::{ToSocketAddrs, UdpSocket};
+
+const SERVICE_PORT: u16 = 6480;
+const LOCAL_PORT_RANGE: RangeInclusive<u16> = 49152..=u16::MAX; // dynamic port range
+const LOCAL_IP: Ipv4Addr = Ipv4Addr::new(10, 5, 0, 2);
+const REMOTE_IP: Ipv4Addr = Ipv4Addr::new(10, 5, 0, 1);
+const HANDSHAKE_VERSION: u32 = 1;
+const CIPHERTEXT_LEN: u32 = kyber768::ciphertext_bytes() as _;
+
+const IPV4_HEADER_LEN: usize = 20;
+const UDP_HEADER_LEN: usize = 8;
+
+/// The PQ module error type
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// IO error
+    #[error("IO: {0:?}")]
+    Io(#[from] io::Error),
+    /// Generic unrecoverable error
+    #[error("Generic: {0}")]
+    Generic(String),
+}
+
+impl From<String> for Error {
+    fn from(value: String) -> Self {
+        Self::Generic(value)
+    }
+}
+
+impl From<&str> for Error {
+    fn from(value: &str) -> Self {
+        Self::Generic(value.to_string())
+    }
+}
+
+impl From<tokio::time::error::Elapsed> for Error {
+    fn from(value: tokio::time::error::Elapsed) -> Self {
+        Self::Io(io::Error::new(io::ErrorKind::TimedOut, value))
+    }
+}
+
+/// Post quantum keys retrived from hanshake
+pub struct PqKeys {
+    /// Kyber shared secret
+    pq_shared: kyber768::SharedSecret,
+    /// X25519 secret key
+    wg_secret: x25519_dalek::StaticSecret,
+}
+
+impl PqKeys {
+    /// Retrieve the PQ preshared key
+    pub fn pq_shared(&self) -> telio_crypto::PresharedKey {
+        #[allow(clippy::expect_used)]
+        telio_crypto::PresharedKey(
+            self.pq_shared
+                .as_bytes()
+                .try_into()
+                .expect("Preshared WG keys should have the same lenght as PQ preshared key"),
```

Comment:
Ehh, so we generally just don't use `expect` in Telio and this doesn't look like some very good place to make an exception, constructor should just return a result IMHO. Also, it looks that it is +/- implementation of `TryFrom` trait, maybe you could turn it into one?

Reply:
I'm afraid I sadly have to disagree, both the WG preshared key and Kyber shared secret are by definition 32 bytes long. The `except` is here only because `as_bytes()` return `&[u8]` instead of `&[u8;32]` which is it's internal representation.
I find creating yet another type of error type and handling logic for errors that will never happen more cumbersome and unnecessary than simply unwrapping

Reply:
Is it used anywhere as `kyber768::SharedSecret` key? Maybe we could just keep it stored as `telio_crypto::PresharedKey`?

Reply:
I have to agree with @stalowyjez this does introduce a sneaky hidden dependency. Can we somehow ensure, that this dependency (e.g. 32 bytes of length) is checked during compile time, so in a situation where this hiddend depenedncy of 32 bytes length is not true anymore -> we would be notified by failing build, instead of runtime crash?

Reply:
Let me try, I guess I will create a custom key type for this one

Reply:
I did move this conversion into the `handshake()` function, returning io error instead of `expect()`